### ISSUE TITLE
i3 says "could not create graphical context" with TigerVNC

### ISSUE THREAD

**Issue by sphaugh** [i3/i3#2729]
Output of `i3 --moreversion 2>&- || i3 --version`:

`Binary i3 version:  4.13 (2016-11-08) © 2009 Michael Stapelberg and contributors`

URL to a logfile as per http://i3wm.org/docs/debugging.html:

http://logs.i3wm.org/logs/5637337312657408.bz2

**What I did:**

Install `tigervnc`, then do
```
cat > ~/.vnc/xstartup <<EOF
#!/bin/sh
exec i3
EOF
chmod +x ~/.vnc/xstartup
vncserver
```

**What I saw:**

```
sean@loefah ~> cat .vnc/loefah:1.log                                                                      

Xvnc TigerVNC 1.7.80 - built Jan  1 1970 00:00:01
Copyright (C) 1999-2016 TigerVNC Team and many others (see README.txt)
See http://www.tigervnc.org for information on TigerVNC.
Underlying X server release 11902000, The X.Org Foundation


Thu Apr  6 16:20:19 2017
 vncext:      VNC extension running!
 vncext:      Listening for VNC connections on all interface(s), port 5901
 vncext:      created VNC server for screen 0
04/06/2017 04:20:22 PM - [libi3] ERROR: Could not create graphical context. Error code: 9. Please report this bug.
04/06/2017 04:20:22 PM - [libi3] ERROR: Could not create graphical context. Error code: 9. Please report this bug.
04/06/2017 04:20:22 PM - [libi3] ERROR: Could not create graphical context. Error code: 9. Please report this bug.
04/06/2017 04:20:22 PM - [libi3] ERROR: Could not create graphical context. Error code: 9. Please report this bug.
04/06/2017 04:20:22 PM - [libi3] ERROR: Could not create graphical context. Error code: 9. Please report this bug.
04/06/2017 04:20:22 PM - [libi3] ERROR: Could not create graphical context. Error code: 9. Please report this bug.
04/06/2017 04:20:22 PM - [libi3] ERROR: Could not create graphical context. Error code: 9. Please report this bug.
04/06/2017 04:20:22 PM - [libi3] ERROR: Could not create graphical context. Error code: 9. Please report this bug.
04/06/2017 04:20:22 PM - [libi3] ERROR: Could not create graphical context. Error code: 9. Please report this bug.
04/06/2017 04:20:22 PM - [libi3] ERROR: Could not create graphical context. Error code: 9. Please report this bug.
04/06/2017 04:20:22 PM - [libi3] ERROR: Could not create graphical context. Error code: 9. Please report this bug.
04/06/2017 04:20:22 PM - [libi3] ERROR: Could not create graphical context. Error code: 9. Please report this bug.
04/06/2017 04:20:22 PM - [libi3] ERROR: Could not create graphical context. Error code: 9. Please report this bug.
04/06/2017 04:20:22 PM - [libi3] ERROR: Could not create graphical context. Error code: 9. Please report this bug.
04/06/2017 04:20:22 PM - [libi3] ERROR: Could not create graphical context. Error code: 9. Please report this bug.
04/06/2017 04:20:22 PM - [libi3] ERROR: Could not create graphical context. Error code: 9. Please report this bug.
04/06/2017 04:20:22 PM - [libi3] ERROR: Could not create graphical context. Error code: 9. Please report this bug.
04/06/2017 04:20:22 PM - [libi3] ERROR: Could not create graphical context. Error code: 9. Please report this bug.
i3status: trying to auto-detect output_format setting
i3status: auto-detected "i3bar"
Could not grab keyboard, status = 3
[libi3] libi3/font.c Using Pango font monospace, size 8
[libi3] libi3/font.c X11 root window dictates 96.094581 DPI
[libi3] libi3/font.c Using Pango font monospace, size 8
```

**What I expected instead:**
A working VNC session using i3.

**Comment by sphaugh** [comment]
It looks like this is NixOS-specific...

**Comment by nmschulte** [comment]
Just adding some diagnostic details.

The error is logged from here: https://github.com/i3/i3/blob/next/libi3/draw_util.c#L48
Error code 9 is `BadDrawable`.

**Comment by sphaugh** [comment]
If I use the default xstartup with twm, I can at least launch xterm and qutebrowser over VNC. I'm not sure what all i3 expects in its environment and if NixOS might be missing something there, though.

**Comment by nmschulte** [comment]
@sphaugh, are you able to build i3 from source?

**Comment by sphaugh** [comment]
Awesome works over VNC same as dwm, and I was able to start an i3 session with `x0vncserver` from the physical machine.

Arch environment, works with vncserver locally and remotely: https://gist.github.com/bf55e9f494741e59b7cb0ea639bd7fe7
NixOS remote environment, can't `x0vncserver` or `vncserver`:
https://gist.github.com/48adc43d4fb1deed53c2c17c1adbc4a2
NixOS physical environment, can at least `x0vncserver`:
https://gist.github.com/a3fd4eda08d2ac843ba99fece644d59e

**Comment by sphaugh** [comment]
@nmschulte I tried building from `next` and didn't see your error message.

**Comment by sphaugh** [comment]
This should build from `next`:

```
diff --git a/pkgs/applications/window-managers/i3/default.nix b/pkgs/applications/window-managers/i3/default.nix
index fded295b0e..e79c543be7 100644
--- a/pkgs/applications/window-managers/i3/default.nix
+++ b/pkgs/applications/window-managers/i3/default.nix
@@ -1,18 +1,19 @@
-{ fetchurl, stdenv, which, pkgconfig, makeWrapper, libxcb, xcbutilkeysyms
-, xcbutil, xcbutilwm, xcbutilxrm, libstartup_notification, libX11, pcre, libev
-, yajl, xcb-util-cursor, coreutils, perl, pango, perlPackages, libxkbcommon
-, xorgserver, xvfb_run, dmenu, i3status }:
+{ fetchgit, stdenv, which, pkgconfig, makeWrapper, autoconf, automake, libtool,
+libxcb, xcbutilkeysyms , xcbutil, xcbutilwm, xcbutilxrm,
+libstartup_notification, libX11, pcre, libev , yajl, xcb-util-cursor,
+coreutils, perl, pango, perlPackages, libxkbcommon , xorgserver, xvfb_run,
+dmenu, i3status }:
 
 stdenv.mkDerivation rec {
   name = "i3-${version}";
   version = "4.13";
 
-  src = fetchurl {
-    url = "http://i3wm.org/downloads/${name}.tar.bz2";
-    sha256 = "12ngz32swh9n85xy0cz1lq16aqi9ys5hq19v589q9a97wn1k3hcl";
+  src = fetchgit {
+    url = https://github.com/i3/i3.git;
+    rev = "refs/heads/next";
   };
 
-  nativeBuildInputs = [ which pkgconfig makeWrapper ];
+  nativeBuildInputs = [ which pkgconfig makeWrapper autoconf automake libtool ];
 
   buildInputs = [
     libxcb xcbutilkeysyms xcbutil xcbutilwm xcbutilxrm libxkbcommon
@@ -22,6 +23,8 @@ stdenv.mkDerivation rec {
     xorgserver xvfb_run
   ];
 
+  preConfigure = "autoreconf -vfi";
+
   configureFlags = [ "--disable-builddir" ];
 
   enableParallelBuilding = true;
```

Maybe take a look at https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/window-managers/i3/default.nix and see if anything stands out?

**Comment by nmschulte** [comment]
@Airblader, I wonder if a call to `xcb_request_check` is expensive or such and maybe it was left out of the `create_window` path intentionally.  It seems the `xcb_create_window` call is not the issue here.  This is over my head at this point.

**Comment by Airblader** [comment]
I don't think the call is expensive like that; it's certainly not the reason we left it out. One thing that makes VNCs »special« is their usually limited support of X extensions. Why the issue only comes up with nixOS, though – I don't know, either.

**Comment by sphaugh** [comment]
What makes i3 different from other window managers like twm, dwm, and awesome, though? Because they all seem fine with VNC.

**Comment by sphaugh** [comment]
Tried starting `vncserver` as root, saw this:
```
Exiting due to signal.
Warning: VNC extension does not support -reset, terminating instead. Use -noreset to prevent termination.
[libi3] libi3/font.c Using Pango font monospace, size 8
[libi3] libi3/font.c X11 root window dictates 96.094581 DPI
```

And this is what I see starting i3 from a bare xterm:
![i3-vnc](https://cloud.githubusercontent.com/assets/8660255/24966371/5e2f68cc-1f6c-11e7-9000-5f4a5505b360.png)

**Comment by psychon** [comment]
My usual shotgun-debugging approach would be: Can you run i3 under [XTrace](https://xtrace.alioth.debian.org/)? For example, instead of `exec i3` just do `exec xtrace i3` (or `x11trace` if that's what the tool is called for you...)

> http://logs.i3wm.org/logs/5637337312657408.bz2

Did someone look at the log file yet? There are other errors showing up in there:
```
04/06/2017 04:15:08 PM - main.c:xcb_check_cb:123 - X11 Error received (probably harmless)! sequence 0x293, error_code = 8
04/06/2017 04:15:08 PM - main.c:xcb_check_cb:123 - X11 Error received (probably harmless)! sequence 0x294, error_code = 3
```
Error 8 is BadMatch and 3 is BadWindow. Also, the difference of the sequence numbers here is just 1, so I guess the first error is an error while CreateWindow and the second one is then a following request that tried to use that window.
(Yes, these errors occur only after some GCs already errored out, but this is still interesting, even though a BadMatch can have many different reasons...)

**Comment by sphaugh** [comment]
@psychon xtrace didn't have any output, sorry. 

```
sean@kodenine ~/.vnc> cat xstartup
#!/bin/sh
exec xtrace /home/sean/.nix-profile/bin/i3

sean@kodenine ~/.vnc> cat kodenine:1.log

Xvnc TigerVNC 1.7.80 - built Jan  1 1970 00:00:01
Copyright (C) 1999-2016 TigerVNC Team and many others (see README.txt)
See http://www.tigervnc.org for information on TigerVNC.
Underlying X server release 11902000, The X.Org Foundation


Thu Jun 15 16:44:08 2017
 vncext:      VNC extension running!
 vncext:      Listening for VNC connections on all interface(s), port 5901
 vncext:      created VNC server for screen 0
Function              File                                                  Line
--------------------------------------------------------------------------------
/run/current-system/sw/bin/xterm: cannot load font '-misc-fixed-medium-r-semicondensed--13-120-75-75-c-60-iso10646-1'
sean@kodenine ~/.vnc>
```

I tried playing with it for a little bit. I can open st and it works after I mouse click the screen, for some reason. But all other functional behavior of i3 is still borked.

**Comment by sphaugh** [comment]
https://gist.github.com/91104206d4dcf51d816fbb969efc2bc5

**Comment by stapelberg** [comment]
Thanks for the log.

Here’s the part where the error message in question originates:
```
000:<:01e8: 52: Request(1): CreateWindow depth=0x18 window=0x0040000c parent=0x00000267 x=-15 y=-15 width=10 height=10 border-width=0 class=InputOutput(0x0001) visual=0x00000178 value-list={background-pixel=0x00000000 border-pixel=0x00000000 override-redirect=true(0x01) event-mask=ButtonPress,ButtonRelease,PointerMotion,Exposure,StructureNotify,SubstructureNotify,SubstructureRedirect colormap=0x00000020}
000:<:01e9: 16: Request(2): ChangeWindowAttributes window=0x0040000c value-list={cursor=0x00400002}
000:<:01ea: 16: Request(55): CreateGC cid=0x0040000d drawable=0x0040000c values={}
000:<:01eb:  4: Request(43): GetInputFocus 
000:>:01e8:Error 8=Match: major=1, minor=0, bad=32
000:>:01e9:Error 3=Window: major=2, minor=0, bad=4194316
000:>:01ea:Error 9=Drawable: major=55, minor=0, bad=4194316
```

As per the request serial number, it is indeed the CreateWindow which fails. I think the `bad=32` refers to the `colormap=0x00000020` (0x20 == 32) which we’re setting.

Issue #2435 was the last time we talked about colormaps, I think.

Could you provide the output of `xwininfo` and `xwininfo -frame` on an xterm in one of the working window managers under your TigerVNC setup please? I’m interested in the “Colormap:” value which should be present in the output.

**Comment by sphaugh** [comment]
dwm: https://gist.github.com/01d7dc9289991b4b5503cd8427049137
https://gist.github.com/af34f39a5499ec3827d24562a2bc53f7

i3: https://gist.github.com/ff2ffa1144bdf641e9118c3683363a03
https://gist.github.com/3dde7bdccfe1b6c9d6105e46171ac3ba

**Comment by stapelberg** [comment]
Okay, so the colormap is 0x20 in either case, but the CreateWindow in your dwm xtrace doesn’t specify the colormap explicitly.

AFAIR, dwm is not actually a reparenting window manager, so could you post the same logs with awesome, please?

Also, do things start working once you apply the following patch to i3?

```diff
diff --git i/src/x.c w/src/x.c
index 09a60493..bfc48059 100644
--- i/src/x.c
+++ w/src/x.c
@@ -142,8 +142,8 @@ void x_con_init(Con *con) {
     mask |= XCB_CW_EVENT_MASK;
     values[3] = FRAME_EVENT_MASK & ~XCB_EVENT_MASK_ENTER_WINDOW;
 
-    mask |= XCB_CW_COLORMAP;
-    values[4] = win_colormap;
+    // mask |= XCB_CW_COLORMAP;
+    // values[4] = win_colormap;
 
     Rect dims = {-15, -15, 10, 10};
     xcb_window_t frame_id = create_window(conn, dims, con->depth, visual, XCB_WINDOW_CLASS_INPUT_OUTPUT, XCURSOR_CURSOR_POINTER, false, mask, values);

```

**Comment by sphaugh** [comment]
awesome: https://gist.github.com/e3a3447a5278dc0aca12a824e0865d5a
https://gist.github.com/dbe337c02cfe0481c842cede257c2958

**Comment by sphaugh** [comment]
I tried the patch and it didn't have any effect. Think I'm out of gists but I can try again later if you need logs.

**Comment by psychon** [comment]
@sphaugh (If you have not already done so), could you provide `xdpyinfo` output? The WM does not matter here. That's hopefully a bit easier to read than the xtrace output and perhaps xtrace is leaving something out (I'm confused about its output, see below).

A working `CreateWindow` from awesome:

000:<:044a: 56: Request(1): CreateWindow depth=0x18 window=0x0040001a parent=0x00000267 x=0 y=0 width=1 height=1 border-width=0 class=CopyFromParent(0x0000) visual=0x00000021 value-list={border-pixel=0x00000000 bit-gravity=NorthWest(0x01) override-redirect=true(0x01) event-mask=ButtonPress,ButtonRelease,EnterWindow,LeaveWindow,PointerMotion,Exposure,StructureNotify,SubstructureNotify,SubstructureRedirect,PropertyChange colormap=0x00000020 cursor=0x00400008}

A non-working `CreateWindow` from i3 which generates a Match error for the colormap:

000:<:02a3: 52: Request(1): CreateWindow depth=0x18 window=0x00400030 parent=0x00000267 x=-15 y=-15 width=10 height=10 border-width=0 class=InputOutput(0x0001) visual=0x00000178 value-list={background-pixel=0x00000000 border-pixel=0x00000000 override-redirect=true(0x01) event-mask=ButtonPress,ButtonRelease,PointerMotion,Exposure,StructureNotify,SubstructureNotify,SubstructureRedirect colormap=0x00000020}

Awesome seems to use a different visual and that's the only (relevant) difference...

awesome's visual:
{depth=24 visuals={id=0x00000021 class=TrueColor(0x04) bits/rgb-value=8 colormap-entries=256 red-mask=0x00ff0000 green-mask=0x0000ff00 blue-mask=0x000000ff};

i3's visual:
{depth=24 visuals={id=0x00000178 class=TrueColor(0x04) bits/rgb-value=8 colormap-entries=256 red-mask=0x00ff0000 green-mask=0x0000ff00 blue-mask=0x000000ff}

Looks identical to me?!? But why does depth=24 appear twice in the list of allowed depths of the root window? And why does this server have so many apparently identical visuals?

**Comment by sphaugh** [comment]
xdpyinfo: https://gist.github.com/ce330f695e7ad5e51371319df516a026

**Comment by psychon** [comment]
Weird...

The code handling CreateWindow is here: https://cgit.freedesktop.org/xorg/xserver/tree/dix/window.c?id=f23e65f96365706c69fa781b2c6cbf3203619c9f#n761
The possible causes of BadMatch in there are:
- non-InputOnly child of an InputOnly window
- InputOnly window with depth or border width
- Window not using the visual or depth of its parent, and not using a visual that is available on the given screen (not applicable here since we have only one screen and that one clearly has the used visuals, right?)
- Trying to use no border on a window with output and a different depth than its parent
- Having no colormap for a window with output that either has a different visual than its parent or where the parent has no colormap

None of this applies to the case we are looking at, I think.

That leaves non-hardcoded error returns.
- The "security hook" can return any error it wants. Does anyone use that?!?
- ChangeWindowAttributes can fail with any error. However, i3 only sets a subset of the options that awesome sets, so why would it fail?

**Comment by psychon** [comment]
OHHHH!!!!! I found it.

Both awesome and i3 use the default colormap, 0x20. However, there is this check:
https://cgit.freedesktop.org/xorg/xserver/tree/dix/window.c?id=f23e65f96365706c69fa781b2c6cbf3203619c9f#n1439
(A colormap can only be used on windows that have the same visual as the colormap)

So since the default colormap, 0x20, is created for the default visual, 0x21, the two can only be used together. However, i3 tries to use colormap 0x20 with visual 0x178. It does not matter that the two visuals are identical, the colormap 0x20 may only be used with visual 0x21.

**Comment by psychon** [comment]
And I guess it's this check in i3 which is wrong:

https://github.com/i3/i3/blob/55bc6741796e8b179b6111a721a3e9631934bb86/src/x.c#L116

This uses `get_visualid_by_depth` to find *any* visual with the given depth. Then it assumes that for any window having the default depth, it can use the visual that was found. However, this really should check if the window has the default visual, not the default depth.

In the case we are investigating, there are two identical visuals (no idea why) and the default visual is later in the list than its duplicate. Thus, `get_visualid_by_depth(24)` will always find this other visual instead of the default visual.

I'll leave writing a patch to someone actually involved with i3. Feel free to acknowledge that I might have been a tiny bit helpful. ;-)

**Comment by stapelberg** [comment]
Thank you very much for your detective work on this issue, @psychon. It’s much appreciated! Be sure to remind me to offer you a beverage of your choice when we run into each other in person :).

If anyone wants to go ahead with a pull request to address this issue, feel free. Otherwise, I’ll have a look when I find some spare time for this (could be weeks).
